docs/cudf/source/user_guide/copy-on-write.md

```diff
@@ -0,0 +1,171 @@
+(copy-on-write-user-doc)=
+
+# Copy-on-write
+
+Copy-on-write is a memory management strategy that allows multiple cuDF objects containing the same data to refer to the same memory address as long as neither of them modify the underlying data.
+With this approach, any operation that generates an unmodified view of an object (such as copies, slices, or methods like `DataFrame.head`) returns a new object that points to the same memory as the original.
+However, when either the existing or new object is _modified_, a copy of the data is made prior to the modification, ensuring that the changes do not propagate between the two objects.
+This behavior is best understood by looking at the examples below.
+
+The default behaviour in cuDF is for copy-on-write to be disabled, so to use it, one must explicitly opt in by setting a cuDF option. This can be done operation by operation if this is required, but the most common use case is to enable copy-on-write for the duration of a script.
```

Comment:
I think this statement is somewhat misleading, unless I am misunderstanding how COW currently works. While it is true that something like this example is safe:
```
x = cudf.DataFrame(...)
# Do things with x
cudf.set_option(copy_on_write, True)
y = cudf.DataFrame(...)
# Do things with y
```
it has potentially surprising semantics. Operations on `x` won't have COW semantics since it predated the setting, right? In fact, I think that it might be operation dependent; if you trigger a deep copy of `x`, then the newly created buffer will be a CoWBuffer once the option has been set, whereas shallow operations will remain non-COW.

I don't think that we need to change any behavior here, but it is worth noting that behavior could be unexpected and that users should probably be sticking to a single mode for the duration of a script.

Reply:
Updated to reflect this.